src/frontend/src/helpers/websocket_client.ts

```diff
@@ -0,0 +1,469 @@
+  private websocket: WebSocket;
+
+  // this allows WebsocketService to access member functions by string key
+  [key: string]: any;
+
+  constructor(private uri: string, private key: Uint8Array, private failure: ()=>void, private closes: ()=>void) {
+    this.coder = new Coder(this.key);
+    this.lastRequest = 0;
+    this.authenticated = false;
+    this.failed = false;
+    this.closed = false;
+    this.started = false;
+
+    this.requests[-1] = {
+      callback: this.authenticate, type: null, time: null
+    };
+    this.requests[-3] = new Request();
+    this.requests[-4] = new Request();
+
+    this.ready = new Promise((resolve, reject)=>{
+      this.websocket = new WebSocket(this.uri);
+      this.started = true;
+      this.websocket.onerror = ()=>{
+        this.failed = true;
+        if(!this.authenticated) {
+          reject("Socket closed during authentication!");
+        }
+        return this.closes && this.closes();
+      };
+      this.websocket.onclose = ()=>{
+        this.closed = true;
+        if(!this.authenticated) {
+          reject("Socket closed during authentication!");
+        }
+        return this.closes && this.closes();
+      };
+
+      this.websocket.onmessage = (message: MessageEvent)=> {
+        var readerT = new FileReader();
+        readerT.onloadend = ()=>{
+          var response_string = readerT.result;
+          var response = <Response>(JSON.parse(response_string));
+          // Assume the response holds the message and response.id holds the
+          //  message identifier that corresponds with it
+          //
+          // response.result will hold the result if the API call succeeded,
+          //  error message otherwise.
+          var request = this.requests[response.id];
+
+          if(request.type != 'ping') {
+            var time = new Date();
+            var diff = request.time ? time.getTime() - request.time : -1;
+            console.log("Received response to message with id "+response.id+" after "+diff+" ms.");
+
+            if(!response.success) {
+              console.error(response);
+            } else {
+              console.log(response);
+            }
+          }
+
+          if(response.success && request.resolve) {
+            request.resolve(response.result);
+          } else if(!response.success && request.reject) {
+            request.reject(response.result);
+          }
+
+          if(request.callback) {
+            request.callback(response.success, response.result);
+          }
+          if(response.id >= 0) {
+            delete this.requests[response.id];
+          }
+        };
+        readerT.readAsText(message.data);
+      };
+      resolve(true);
+    });
+  }
+
+  public close(): void {
+    this.websocket.close();
+  }
+
+  private async authenticate(success: boolean, server_challenge: AuthResult): Promise<boolean> {
+    // Generate a nonce
+    var client_nonce = new Uint8Array(sjcl.random.randomWords(32));
```

Comment:
window.getRandomBytes (so we don't have to depend on another JavaScript library)